contracts/exchangeIssuance/FlashMintNotional.sol

```diff
@@ -0,0 +1,901 @@
+        // Return output amount
+        return outputAmount;
+    }
+
+
+    /**
+     * Sells all components (after redemption of fCash positions) for the output token
+     */
+    function _sellComponentsForOutputToken(
+        TradeData memory _tradeData,
+        DEXAdapter.SwapData[] memory _swapData
+    )
+        internal
+    {
+        (address[] memory components, uint256[] memory componentUnits) = _getFilteredComponentsRedemption(
+            _tradeData.setToken,
+            _tradeData.amountSetToken,
+            _tradeData.issuanceModule,
+            _tradeData.isDebtIssuance,
+            _tradeData.slippage
+        );
+
+        require(components.length == _swapData.length, "Components / Swapdata mismatch");
+
+        for (uint256 i = 0; i < components.length; i++) {
+            uint256 maxAmountSell = componentUnits[i];
+            address component = components[i];
+            if(component == address(0)){
+                break;
+            }
+
+            // If the component is equal to the output token we don't have to trade
+            if(component != address(_tradeData.paymentToken)) {
+                addresses.swapExactTokensForTokens(maxAmountSell, 0, _swapData[i]);
+            }
+
+        }
+    }
+
+    /**
+     * Redeem all fCash positions for the underlying token
+     */
+    function _redeemWrappedFCashPositions(
+        TradeData memory _tradeData
+    ) 
+    internal
+    {
+        (address[] memory components, uint256[] memory componentUnits) = getRequiredRedemptionComponents(
+            _tradeData.issuanceModule,
+            _tradeData.isDebtIssuance,
+            _tradeData.setToken,
+            _tradeData.amountSetToken
+        );
+
+        for (uint256 i = 0; i < components.length; i++) {
+            address component = components[i];
+            uint256 units = componentUnits[i];
+
+            if(_isWrappedFCash(component)) {
+                IWrappedfCash(component).redeemToUnderlying(units, address(this), 0);
+            }
+        }
+    }
+
+    /**
+     * Mint all fCash positions from the underlying token
+     */
+    function _mintWrappedFCashPositions(
+        TradeData memory _tradeData,
+        address[] memory componentsBought,
+        uint256[] memory amountsAvailable
+    ) 
+    internal
+    {
+        (address[] memory components, uint256[] memory componentUnits) = getRequiredIssuanceComponents(
+            _tradeData.issuanceModule,
+            _tradeData.isDebtIssuance,
+            _tradeData.setToken,
+            _tradeData.amountSetToken
+        );
+
+        for (uint256 i = 0; i < components.length; i++) {
+            address component = components[i];
+            uint256 units = componentUnits[i];
+            if(_isWrappedFCash(component)) {
+                IERC20 underlyingToken = _getUnderlyingToken(IWrappedfCash(component));
+                uint256 componentIndex = _findComponent(componentsBought, address(underlyingToken));
```

Comment:
I think we would get an array out of bounds exception in the next line.
However I don't see a scenario where that would happen since this list of bought components was generated in this contract, in the same transaction from the `_getFilteredComponents` method via the `_buyComponentsForIssuance` method.
This is *not* data that the user can control as far as I understand.

Reply:
Thinking about this now, there might be gas optimization potential here though if we return mapping of component index -> filteredComponent index from the `_getFilteredComponents` method so we don't have to search every component here again.

Reply:
I implemented an alternative that uses a cached mapping-array here instead of calling `_findComponent` again.
Let me know what you think (especially regarding wether this improves readability / complexity):
https://github.com/IndexCoop/index-coop-smart-contracts/pull/105/commits/2920135cf0b71c46bbdb0eb8189d6ba1bf089a00

I didn't actually evaluate the gas difference but I don't expect it to be major.